### ISSUE TITLE
Intermittent ERROR: Couldn't connect to Docker daemon at http+docker://localunixsocket - is it running?

### ISSUE THREAD

**Issue by piotrb** [buildkite/elastic-ci-stack-for-aws#236]
Having a bit of an odd random occuring issue ..

setup is elastic cloud 2.rc1-9, beta agent, docker compose

occasionally .. steps fail to connect to docker with:

```
Creating network "buildkite5736e56df5c2478db2635348bf21e4d2_default" with the default driver
ERROR: Couldn't connect to Docker daemon at http+docker://localunixsocket - is it running?
```

except that .. earlier in that step it was able to docker pull just fine .. and later in that step in cleanup its able to try to clean up without connection errors (it said no network to delete, etc etc, which means it connected and got inventory right?) ..

Not sure if this is a elastic-ci issue, or agent, or plugin or what ..

**Comment by piotrb** [comment]
This is quite annoying since its causing random failures ..

**Comment by lox** [comment]
I'm sorry you've been having issues, we will investigate. I believe
@bjeanes might have had similar issues.

**Comment by keithpitt** [comment]
@piotrb whats your agent count / server setup like? Are you running lots of agents on 1 machine? Or 1 agent per machine?

**Comment by bjeanes** [comment]
I get this occasionally. One of my stacks is 4 agents/server but I can't recall if this error was isolated to that stacks or not.

**Comment by piotrb** [comment]
Running 4 agents per m4.xlarge instance

**Comment by kbaum** [comment]
I also get this error when running 4 docker instances on a t2.xlarge.

**Comment by lox** [comment]
Any idea if these instances are scaling down when they error?

**Comment by piotrb** [comment]
I don't know if I can ever really catch that .. 

Oh yea .. I guess I havn't seen this issue in a while now .. so somehing has changed

Although it seems like docker comes back up since it tries do do the cleanup steps .. so its not like its shut down already or something

**Comment by kbaum** [comment]
In my situation, the instances were just launched.

**Comment by lox** [comment]
That makes sense too, it would happen either before docker starts or after
it shuts down. I will do some research.

**Comment by kbaum** [comment]
Cut down to 4 docker instances on t2.xlarge ec2 instances but still regularly seeing this error.  Any known workarounds?

**Comment by lox** [comment]
Hrm, this could be from a variety of things. I'd love to see the docker log after this happens. Any chance you'd have a look on the next failure @kbaum? Would like to find a solution to this for you folks.

**Comment by kbaum** [comment]
@lox - sending docker logs over slack to you.  Thx!

**Comment by lox** [comment]
Interesting, from those logs it looks like it's not what I'd initially thought, which was that the docker daemon had started or stopped, but it looks like it's failing half way through an operation. 

```
time="2017-02-08T02:13:06.538884475Z" level=error msg="attach: stdout: write unix /var/run/docker.sock->@: write: broken pipe"
```

**Comment by lox** [comment]
Ok, I think it's https://github.com/docker/docker/issues/27226. Fixed in Docker 1.12.6.

**Comment by kbaum** [comment]
@lox - Upgraded to the Buildkite stack v2.0.0-rc3-1-g331eb6c stack and ran into the same issue:

https://gist.github.com/913ca009a497a774565f28673de4cfbc

**Comment by lox** [comment]
Ugh. Docker. Ok, sorry for this @kbaum, I'll file a docker bug and see where we get.

**Comment by kbaum** [comment]
Thanks @lox.  It seems like I am running into this issue more than others.  Is there something I am doing differently that's causing this bug to be surfaced more often?  At least one ec2 instance on the majority of the builds is running into this problem so can't really hand this off to entire team yet.

**Comment by piotrb** [comment]
A bit more info about an instance of this just now .. 

The instance has already been running for 2 hours, it is NOT shutting down (other jobs are running on it)

**Comment by lox** [comment]
The only thing I can think that we removed is a defensive restart of the docker daemon on start. I've re-added that and we can test.

**Comment by lox** [comment]
I've released v2.0.0-rc4. It's at https://s3.amazonaws.com/buildkite-aws-stack/aws-stack.json.

**Comment by piotrb** [comment]
@lox this version doesn't seem to have all the AMIs properly in the mapping table, just us-east-1 in there (we use us-west-2).

**Comment by lox** [comment]
Indeed, apologies @piotrb. Try now, released as v2.0.0-rc5.

**Comment by piotrb** [comment]
On RC5 now and got this issue .. seems new and potentially related .. 

```
> Authenticating with AWS ECR	0s
Flag --email has been deprecated, will be removed in 1.14.
Warning: failed to get default registry endpoint from daemon (Cannot connect to the Docker daemon at unix:///var/run/docker.sock. Is the docker daemon running?). Using system default: https://index.docker.io/v1/
Cannot connect to the Docker daemon at unix:///var/run/docker.sock. Is the docker daemon running?
```

Weird .. didn't see it fail with docker daemon issue before this high up

**Comment by piotrb** [comment]
On another instance got: 

```
> Downloading env files	2s
Waiting for Docker...
Cannot connect to the Docker daemon at unix:///var/run/docker.sock. Is the docker daemon running?
```

This is even higher up ..

**Comment by piotrb** [comment]
Both those seem very intermittent, I think they ran on very fresh instances, looks like workers on one instance got one of those, and workers on another instance got the other. Workers on all other instances ran just fine.

**Comment by piotrb** [comment]
Not sure if this is related (first time seeing it with Rc5 .. 

This is right the first thing in the build, literally nothing has ran .. 

```
> Setting up plugins	3s
# Could not aquire lock on "/var/lib/buildkite-agent/plugins/github-com-buildkite-plugins-docker-compose-buildkite-plugin.lock" (Locked by other process)
# Trying again in 1 second...
# Could not aquire lock on "/var/lib/buildkite-agent/plugins/github-com-buildkite-plugins-docker-compose-buildkite-plugin.lock" (Locked by other process)
# Trying again in 1 second...
# Could not aquire lock on "/var/lib/buildkite-agent/plugins/github-com-buildkite-plugins-docker-compose-buildkite-plugin.lock" (Locked by other process)
# Trying again in 1 second...
🚨 Buildkite Error: open /var/lib/buildkite-agent/plugins/github-com-buildkite-plugins-docker-compose-buildkite-plugin.lock: no such file or directory
```

**Comment by lox** [comment]
Just saw this myself too.

**Comment by lox** [comment]
Have just released 2.0.0-rc6.

**Comment by piotrb** [comment]
Weird .. the template link still says rc5 in its description ...

**Comment by piotrb** [comment]
What changed related to this issue in rc6?

**Comment by piotrb** [comment]
Oh .. https://github.com/buildkite/elastic-ci-stack-for-aws/pull/257 looks interesting ... could just be it ...

**Comment by deppy** [comment]
So this issue doesn't seem to be fixed by #257 but I have some more details.  It still seems to happen shortly after a new instance is started, but takes about 10 minutes from the instance start time until the issue shows up.  I'm seeing the job output just randomly stop and I can't ssh into the instance.  So I ended up rebooting one of the instances that this appeared on and I pulled all the logs from it (Can provide any/all as requested).

The thing that stands out the most is in the docker logs:  https://gist.github.com/deppy/d15dec4d3ef0ecfd5704760e3df9bc1b

I was also already on an instance when this presented itself and I was barely able to run any commands, but I could confirm that docker thought the containers were still running and I wasn't able to `docker exec` to any of them.

In one of the cases today the instance stayed connected to buildkite and was accepting jobs and they would fail (this was the one I was already connected to).  The other case today the instance was disconnected due to timeout (this was the one I got the logs from).

**Comment by lox** [comment]
Ugh, that's a bummer @deppy. That would line up with my theory that something was causing containerd to become non-responsive. 

@kbaum is running 2.0.0-rc1 (https://s3.amazonaws.com/buildkite-aws-stack/master/7ddb893abf326592a8a6bf7053891d8af991544f.aws-stack.json) without issue. Could I trouble you to update to that and see if the problem persists? It would certainly be useful if I could figure out when the last known good stack was so I could bisect down to the issue.

@piotrb how is rc6 running for you?

**Comment by lox** [comment]
This is exactly the issue: https://superuser.com/questions/1167873/docker-daemon-hanging-missing-container-state-file

**Comment by lox** [comment]
Also, this https://github.com/docker/docker/issues/27211#issuecomment-285546593

**Comment by lox** [comment]
Ok, I'm going to close this since I've released a fix for it in 2.0.2 and we can open new more specific one if we need, as I believe this bug is actually caused by multiple things.
